python/dgl/dataloading2/graph_storage/csc_sampling_graph.py

```diff
@@ -0,0 +1,308 @@
+            self.c_csc_graph.type_per_edge()
+            if self.c_csc_graph.is_heterogeneous()
+            else None
+        )
+
+
+def from_csc(
+    csc_indptr: torch.Tensor,
+    indices: torch.Tensor,
+    etype_sorted: bool = True,
+    num_nodes: Optional[int] = None,
+    hetero_info: Optional[HeteroInfo] = None,
+) -> CSCSamplingGraph:
+    """
+    Create a CSCSamplingGraph object from a CSC representation.
+
+    Parameters
+    ----------
+    csc_indptr : torch.Tensor
+        Pointer to the start of each row in the `indices`.
+    indices : torch.Tensor
+        Column indices of the non-zero elements in the CSC graph.
+    etype_sorted : bool
+        A hint telling whether neighbors of each node are already sorted by edge type ID,
+        it is only meaningful when graph is heterogeneous, by default True.
+    num_nodes : int, optional
+        Number of nodes in the graph, by default None.
+    hetero_info : Optional[HeteroInfo], optional
+        Heterogeneous graph metadata, by default None.
+        Indicates whether the graph is homogeneous or heterogeneous.
+
+    Returns
+    -------
+    CSCSamplingGraph
+        The created CSCSamplingGraph object.
+
+    Examples
+    --------
+    >>> csc_indptr = torch.tensor([0, 2, 5, 7])
+    >>> indices = torch.tensor([1, 3, 0, 1, 2, 0, 3])
+    >>> num_nodes = 3
+    >>> ntypes, etypes = ['n1', 'n2'], ['e1', 'e2']
+    >>> hetero_info = (ntypes, etypes, torch.tensor([0, 2]), torch.tensor([0, 1, 0, 1, 1, 0, 0]))
+    >>> graph = from_csc(csc_indptr, indices, num_nodes, hetero_info)
+    >>> print(graph)
+    CSCSamplingGraph(csc_indptr=tensor([0, 2, 5, 7]),
+                    indices=tensor([1, 3, 0, 1, 2, 0, 3]),
+                    num_nodes=3, num_edges=7, is_heterogeneous=True)
+    """
+    if num_nodes is None:
+        num_nodes = csc_indptr.shape[0] - 1
+
+    assert csc_indptr.shape[0] == num_nodes + 1
+    assert torch.max(indices) < num_nodes
+    if hetero_info is None:
+        return CSCSamplingGraph(
+            torch.ops.graphbolt.from_csc(num_nodes, csc_indptr, indices)
+        )
+    else:
+        if etype_sorted is False:
+            ntypes, etypes, node_type_offset, type_per_edge = hetero_info
+            # sort neighbors by edge type id for each node
+            sorted_etype = []
+            sorted_indices = []
+            for s, e in zip(csc_indptr[:-1], csc_indptr[1:]):
+                neighbor_etype, index = torch.sort(type_per_edge[s:e])
+                sorted_etype.append(neighbor_etype)
+                sorted_indices.append(indices[s:e][index])
+            type_per_edge = torch.cat(sorted_etype, dim=0)
+            indices = torch.cat(sorted_indices, dim=0)
+            hetero_info = (ntypes, etypes, node_type_offset, type_per_edge)
+
+        return CSCSamplingGraph(
+            torch.ops.graphbolt.from_csc_with_hetero_info(
+                num_nodes, csc_indptr, indices, *hetero_info
+            )
+        )
+
+
+def from_coo(
```

Comment:
From API usability perspective:
Is this from_coo better or from_coo with map from `canonical edge type` to `coo torch.Tensor` better?